app/src/main/java/com/keylesspalace/tusky/db/DraftsAlert.kt

```diff
@@ -0,0 +1,126 @@
+
+    fun updateActiveAccountId(dbId: Long) {
+        if (dbId == userIdCurrent) {
+            return; // Nothing to do
+        }
+
+        userIdCurrent = dbId
+
+        // Remove old observer if we just switched accounts
+        draftsNeedUserAlertCurrent?.let {
+            draftsNeedUserAlert.removeSource(it)
+        }
+
+        // Add new observer
+        val draftsNeedUserAlertCurrent = draftDao.draftsNeedUserAlert(dbId)
+        this.draftsNeedUserAlertCurrent = draftsNeedUserAlertCurrent
+        draftsNeedUserAlert.addSource(draftsNeedUserAlertCurrent, { count ->
+            Log.d(TAG, "draftsNeedUserAlert: account id " + dbId + " has " + count + " notification-worthy drafts") // REMOVE BEFORE COMMIT
+            // !! here is safe because this is called only after userIdCurrent first goes non-null, and it never goes null again
+            draftsNeedUserAlert.value = Pair(userIdCurrent!!, count)
+        })
+    }
+
+    init {
+        accountManager.draftsAlert = this
+        accountManager.activeAccount?.let { updateActiveAccountId(it.id) }
+    }
+
+    public fun <T> observeInContext(context: T, showAlert: Boolean) where T : Context, T : LifecycleOwner {
+        // observe ensures that this gets called at the most appropriate moment wrt the context lifecycle—
+        // at init, at next onResume, or immediately if the context is resumed already.
+        if (showAlert) {
+            draftsNeedUserAlert.observe(context) { (dbId, count) ->
+                Log.d(TAG, "draftsNeedUserAlert 2: Notification-worthy draft count " + count)
+                if (count > 0) {
+                    AlertDialog.Builder(context)
+                        .setTitle(R.string.action_post_failed)
+                        .setMessage(
+                            context.getResources().getQuantityString(R.plurals.action_post_failed_detail, count)
+                        )
+                        .setPositiveButton(R.string.action_post_failed_show_drafts) { _: DialogInterface?, _: Int ->
+                            clearDraftsAlert(dbId) // User looked at drafts
+
+                            val intent = DraftsActivity.newIntent(context)
+                            context.startActivity(intent)
+                        }
+                        .setNegativeButton(R.string.action_post_failed_do_nothing) { _: DialogInterface?, _: Int ->
+                            clearDraftsAlert(dbId) // User doesn't care
+                        }
+                        .show()
+                }
+            }
+        } else {
+            draftsNeedUserAlert.observe(context) { (dbId, _) ->
+                Log.d(TAG, "draftsNeedUserAlert 3: Clean out")
+                clearDraftsAlert(dbId)
+            }
+        }
+    }
+
+    /**
+     * Clear drafts alert for current user
+     * Caller's responsibility to ensure there is a current user
+     */
+    fun clearDraftsAlert(id: Long) {
+        GlobalScope.launch {

```

Comment:
Not crazy about this but not sure how else to do it. Could I get a coroutine scope from a lifecycle? In that case I could pass it in from the calls above, because this is always invoked in the context of the alert box and we always have a particular LifecycleOwner that is called with respect to. Would that allow this coroutine to continue to execute even after that lifecycle suspends (which if we displayed the dialog, will always be right after we do this), though?

What we need is for this to execute before any further draftDao events get routed in to the LiveData mediator.

Reply:
Yes this needs to use the scope of the activity. You can get it by calling `lifecycleScope` on an Activity/LifecycleOwner.

Reply:
Made this change, what happens tho if the LifecycleOwner suspends while the launched async task is still unfinished? Like is there a potential for a race condition or are launched jobs allowed to finish?